plugin/src/main/groovy/com/novoda/staticanalysis/internal/SourceFilter.groovy

```diff
@@ -0,0 +1,43 @@
+package com.novoda.staticanalysis.internal
+
+import org.gradle.api.Project
+import org.gradle.api.file.FileCollection
+import org.gradle.api.file.FileTree
+import org.gradle.api.tasks.SourceTask
+
+class SourceFilter {
+
+    private final Project project
+    private final List<Object> excludes = []
+
+    SourceFilter(Project project) {
+        this.project = project
+    }
+
+    void exclude(Object exclude) {
+        excludes.add(exclude)
+    }
+
+    void applyTo(SourceTask task) {
+        excludes.each { exclude ->
+            if (exclude instanceof File) {
+                apply(task, project.files(exclude))
+            } else if (exclude instanceof FileCollection) {
+                apply(task, exclude)
+            } else if (exclude instanceof Iterable<File>) {
+                apply(task, exclude.inject(null, accumulateIntoTree()) as FileTree)
```

Comment:
yep. why `inject` instead of `fold` or the more popular `reduce` is beyond me :/